sonarbuild.bat

```diff
@@ -0,0 +1,150 @@
+	if ERRORLEVEL 1 ( 
+		echo Cannot find the build wrapper %SONAR_WRAPPER%
+		set BUILD_WRAPPER=)
+	where /q %SONAR_SCANNER%
+	if ERRORLEVEL 1 (
+		echo Cannot find the Sonar Scanner %SONAR_SCANNER%
+		set BUILD_WRAPPER=)
+	
+	if "%SONAR_KEY%"=="" ( set BUILD_WRAPPER=)
+	if "%SONAR_LOGIN%"=="" ( set BUILD_WRAPPER=)
+	if "%SONAR_ORGANIZATION%"=="" ( set BUILD_WRAPPER=)
+	if "%SONAR_HOST_URL%"=="" ( set BUILD_WRAPPER=)
+	IF "%BUILD_WRAPPER%"=="" (
+		echo Proceeding WITHOUT SonarQube analysis
+	)
+	if NOT "%BUILD_WRAPPER%"=="" (
+		echo Everything looks good. SonarQube analysis is enabled for this build.
+	)
+rem Clean up after ourselves
+	SET shouldsonar=
+	
+:skipsonar
+
+set PATH=%PATH%;%MSBUILD_BIN_PATH%
+
+REM Enable AppVeyor build message logging if running under AppVeyor
+IF "%APPVEYOR%"=="True" set CI_BUILD_ARG="/l:C:\Program Files\AppVeyor\BuildAgent\Appveyor.MSBuildLogger.dll"
+REM Set up SonarQube user-mode analysis if BUILD_WRAPPER environment variable set
+IF NOT "%BUILD_WRAPPER%"=="" (
+	%SONAR_SCANNER% begin /k:"%SONAR_KEY%" /d:sonar.login="%SONAR_LOGIN%" /d:sonar.organization="%SONAR_ORGANIZATION%" /d:sonar.host.url="%SONAR_HOST_URL%" /d:sonar.cfamily.build-wrapper-output=!SONAR_CFAMILY_BUILDWRAPPEROUTPUT!
+)
+REM Build
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Release;Platform=x64 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Release;Platform=Win32 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Release;Platform=ARM /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Release;Platform=ARM64 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Debug;Platform=x64 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Debug;Platform=Win32 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Debug;Platform=ARM /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration=Debug;Platform=ARM64 /t:Rebuild !CI_BUILD_ARG! || goto buildfailed
+goto usermodedone
+
+:buildfailed
+SET failedERROR=%ERRORLEVEL%
+SET buildfailed=true
+
+:usermodedone
+REM End user-space wrapper
+IF NOT "%BUILD_WRAPPER%"=="" (
+	%SONAR_SCANNER% end /d:sonar.login="%SONAR_LOGIN%"
+)
+
+IF "%buildfailed%"=="true" ( goto buildfailed)
+
+
+REM Set up SonarQube kernel-mode analysis if SONAR_KEY_SYS environment variable set
+IF NOT "%BUILD_WRAPPER%"=="" (
+	IF NOT "%SONAR_KEY_SYS%"=="" (
+		%SONAR_SCANNER% begin /k:"%SONAR_KEY_SYS%" /d:sonar.login="%SONAR_LOGIN%" /d:sonar.organization="%SONAR_ORGANIZATION%" /d:sonar.host.url="%SONAR_HOST_URL%" /d:sonar.cfamily.build-wrapper-output=!SONAR_CFAMILY_BUILDWRAPPEROUTPUT!
+	)
+)
+
+%BUILD_WRAPPER% msbuild dokan.sln /p:Configuration="Win10 Release";Platform=x64 /t:Rebuild !CI_BUILD_ARG! || goto buildfailedsys
```

Comment:
I have not cross-checked this, and my batch-skills became a bit rusty over the years: Can we somehow share some code with build.bat?

Reply:
I actually wrote this to do exactly the same as build.bat if it doesn't have the SonarQube variables set, which means "if sonarvars.bat doesn't exist".  (I could rem out the SET commands in sonarvars.bat so that they wouldn't run, but it would make things more difficult if people didn't already understand that they needed to remove the REM statements.  That's why I didn't offer a patch to overwrite build.bat.)

What I did was take build.bat and add a bunch of variable setting and checking.  I also added %BUILD_WRAPPER% invocations before the 'msbuild' lines, because if you don't run it through the build wrapper then SonarScanner doesn't capture the output to figure out what it's supposed to analyze.  (if %BUILD_WRAPPER% isn't set, then it expands to an empty string, so the msbuild command becomes the first thing on the line.)

Theoretically, it might be possible to share code with build.bat.  It would require changing the order of the builds in build.bat (because of SonarScanner's ordering issue and the sys/ directory not being compiled in the non-version-specific msbuilds), and would require some fancy string matching in a FOR loop (so that only lines that start with "msbuild" would be invoked).  It would also be lot more difficult (read: I personally can't figure out a way) to separate out the non-version-specific builds for general analysis and the version-specific builds for sys/ directory analysis.  Input on how to accomplish it would be appreciated, since sys/ directory analysis is a concern.

Reply:
Hi @kyanha,
I agree with @Rondom it is a good idea you have but it share too much code with `build.bat`. We do not need to stick with batch code.
What could be done is to move `build.bat` and `dokan_wix` in scripts folder. Rewrite them in powershell to use helpers (`build_helper.ps1`) to avoid duplicate codes like vswhere / foreach build configuration and platform, .... Even create a coverity build script that use/download official tools instead of coveritypublisher https://github.com/dokan-dev/dokany/issues/665
Like that appveyor script would be smaller and only set the needed env variables for coverity / sonar like you did with `sonarvars.bat` and call our build scripts.